src/extractor/MappingExtractor.ts

```diff
@@ -0,0 +1,101 @@
+import { fhirdefs, fhirtypes, fshtypes, utils } from 'fsh-sushi';
+import { differenceWith, isEqual, pullAt } from 'lodash';
+import { ExportableMapping, ExportableMappingRule } from '../exportable';
+import { ProcessableElementDefinition, ProcessableStructureDefinition } from '../processor';
+import { getPath } from '../utils';
+
+export class MappingExtractor {
+  static process(
+    input: ProcessableStructureDefinition,
+    elements: ProcessableElementDefinition[],
+    fhir: fhirdefs.FHIRDefinitions
+  ): ExportableMapping[] {
+    const mappings =
+      input.mapping?.map(m => {
+        const mapping = new ExportableMapping(m.identity);
+        mapping.source = input.name;
+        if (m.name) mapping.title = m.name;
+        if (m.uri) mapping.target = m.uri;
+        if (m.comment) mapping.description = m.comment;
+        return mapping;
+      }) ?? [];
+    elements.forEach(element => {
+      this.extractRules(element, mappings);
+    });
+
+    // Filter out mappings on the parent - only include mappings new to the profile
+    // or mappings on the base definition with additional mapping rules
+    const parent = fhir.fishForFHIR(
+      input.baseDefinition,
+      utils.Type.Resource,
+      utils.Type.Type,
+      utils.Type.Profile,
+      utils.Type.Extension
+    );
+    const newItems = differenceWith(input.mapping, parent?.mapping, isEqual);
+    const newMappings = mappings.filter(mapping => newItems.some(i => i.identity === mapping.name));
+    const inheritedMappings = mappings.filter(mapping =>
+      parent?.mapping.some((i: fhirtypes.StructureDefinitionMapping) => i.identity === mapping.name)
+    );
+
+    // For each inherited mapping, look to see if there are any new rules
+    const changedInheritedMappings = inheritedMappings.filter(mapping => {
+      const changedIndexes: number[] = [];
+      let isNew = false;
+      mapping.rules.forEach((rule, i) => {
+        const element = fhirtypes.StructureDefinition.fromJSON(parent).findElementByPath(
+          rule.path,
+          fhir
+        );
+        // See if any mapping on the element matches the current rule we are looking for
+        // If one matches, this is not a new rule. If none match, this is new to the profile.
+        const elementMappingRules = element.mapping?.map((m, i) =>
+          this.processMappingRule(ProcessableElementDefinition.fromJSON(element), m, i)
+        );
+        if (!elementMappingRules.find(r => isEqual(r, rule))) {
+          isNew = true;
+        } else {
+          // Remove inherited rules so we don't include them if the mapping is exported
+          changedIndexes.push(i);
+        }
+      });
+      pullAt(mapping.rules, changedIndexes);
+      return isNew;
+    });
```

Comment:
Hm this is an intriguing refactor idea. When I was iterating through the rules of the existing mappings, I was trying to limit how many additional elements I was looking at and creating rules for. But maybe in the end I didn't actually save myself that many iterations, so maybe this would simplify the code enough to make it worth it. I'll take a look (it possibly (probably) will extend to Monday morning though).